File: algorithm/greedy_coin_changing.cpp
#include <iostream>
#include <string>
#include <vector>
#include <math.h>
using namespace std;

void greedyChanges(int cur[], int n, int target, int res[]){
    int max;
    int idx = n - 1;//note idx of current max
    while(target > 0){
        for(int i = idx; i >= 0; i--){
            if(cur[i] <= target){
                max = cur[i];
                idx = i;
                break;
            }
        }
        res[idx]++;
        target -= max;
    }
}

int main(){
    cout<<"input number of types of currency: "<<endl;
    int num;
    cin>>num;

    cout<<"input each type of currency: "<<endl;
    int currency[num];
    for(int i = 0; i < num; i++){
        int money;
        cin>>money;
        currency[i] = money;
    }

    cout<<"input needed changes: "<<endl;
    int target;
    cin>>target;

    int result[num];//initial set to record changes
    for(int i = 0; i < num; i++){
        result[i] = 0;
    }
    greedyChanges(currency,num,target,result);
    cout<<"change result is: "<<endl;
    for(int i = 0; i < num; i++){
        if(result[i] != 0){
            cout<<result[i]<<" numbers of currency "<<currency[i]<<endl;
        }
    }
    return 0;
}

File: algorithm/greedy_coin_changing.js
function greedyChanges(cur, n, target, res){
    let max;
    let idx = n - 1;//note idx of current max
    while(target > 0){
        for(let i = idx; i >= 0; i--){
            if(cur[i] <= target){
                max = cur[i];
                idx = i;
                break;
            }
        }
        res[idx]++;
        target -= max;
    }
}


let num = 8;
let currency = [1,2,5,10,20,50,100,500];
let targets = [13,45,76,89,672];
for(let i = 0; i < targets.length; i++){
	let target = targets[i];
	let res = new Array(8);
	res.fill(0);
	greedyChanges(currency,num,target,res);
	console.log("coin changes for ",target);
	for(let j = 0; j < res.length; j++){
		if(res[j] != 0){
			console.log("numbers of type",currency[j]);
			console.log("is",res[j]);
		}
	}
}
